optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -181,4 +187,57 @@ public boolean isFieldADeepCloneProperty(Field field, Class<?> owningClass) {
         }
         return false;
     }
+
+    /**
+     * @return never null
+     */
+    public Set<Class<?>> getDeepClonedTypeArguments(Type genericType) {
+        // Check the generic type arguments of the field.
+        // It is possible for fields and methods, but not instances.
+        if (!(genericType instanceof ParameterizedType)) {
+            return Collections.emptySet();
+        }
+
+        Set<Class<?>> deepClonedTypeArguments = new HashSet<>();
+        ParameterizedType parameterizedType = (ParameterizedType) genericType;
+        for (Type actualTypeArgument : parameterizedType.getActualTypeArguments()) {
+            if (actualTypeArgument instanceof Class
+                    && isClassDeepCloned((Class) actualTypeArgument)) {
+                deepClonedTypeArguments.add((Class) actualTypeArgument);
+            }
+            deepClonedTypeArguments.addAll(getDeepClonedTypeArguments(actualTypeArgument));
+        }
+        return deepClonedTypeArguments;
+    }
+
+    public Set<Class<?>> getDeepClonedClasses(Collection<Class<?>> entitySubclasses) {
+        Set<Class<?>> deepClonedClassSet = new HashSet<>();
+
+        Stream.of(Stream.of(solutionDescriptor.getSolutionClass()),
+                solutionDescriptor.getEntityClassSet().stream(),
+                entitySubclasses.stream())
+                .flatMap(classStream -> classStream)
```

Comment:
`Stream.concat()` seems easier to read in this case.

Reply:
`Stream.concat()` only exists for two arguments, so it would be
`Stream.concat(Stream.of(solutionDescriptor.getSolutionClass()),
Stream.concat(solutionDescriptor.getEntityClassSet().stream(),entitySubclasses.stream()))` (and if for each argument, another `Stream.concat` is needed). `Stream.of(...).flatmap(s -> s)` works for any number of arguments .